apps/arweave/src/ar_node_worker.erl

```diff
@@ -493,59 +493,110 @@ handle_info({event, miner, {found_solution, Solution, PoACache, PoA2Cache}}, Sta
 				Now
 		end,
 
-	%% Load key
-	RewardKey = case ar_wallet:load_key(MiningAddress) of
-		not_found ->
-			?LOG_WARNING([{event, mined_block_but_no_mining_key_found}, {node, node()},
-					{mining_address, ar_util:encode(MiningAddress)}]),
-			ar:console("WARNING. Can't find key ~w~n", [ar_util:encode(MiningAddress)]),
-			not_found;
-		Key ->
-			Key
-	end,
-	PassesKeyCheck = RewardKey =/= not_found,
-
-	%% Check solution difficulty
-	Diff = get_current_diff(Timestamp),
-	PassesDiffCheck = PassesKeyCheck andalso binary:decode_unsigned(SolutionH, big) > Diff,
+	[{wallet_list, WalletList}] = ets:lookup(node_state, wallet_list),
+	IsBanned = is_account_banned(MiningAddress, ar_wallets:get(WalletList, MiningAddress)),
 
-	%% Check that solution is later than the previous solution on the timeline
+	%% Check the solution is ahead of the previous solution on the timeline.
 	[{_, TipNonceLimiterInfo}] = ets:lookup(node_state, nonce_limiter_info),
 	NonceLimiterInfo = #nonce_limiter_info{ global_step_number = StepNumber,
 			output = NonceLimiterOutput,
 			prev_output = TipNonceLimiterInfo#nonce_limiter_info.output },
-	PassesTimelineCheck = PassesDiffCheck andalso
-			ar_nonce_limiter:is_ahead_on_the_timeline(NonceLimiterInfo, TipNonceLimiterInfo),
+	PassesTimelineCheck =
+		case IsBanned of
+			true ->
+				ar_events:send(solution, {rejected, #{ reason => mining_address_banned,
+						source => Source }}),
+				{false, address_banned};
+			false ->
+				case ar_nonce_limiter:is_ahead_on_the_timeline(NonceLimiterInfo,
+						TipNonceLimiterInfo) of
+					false ->
+						ar_events:send(solution, {stale, #{ source => Source }}),
+						{false, timeline};
+					true ->
+						true
+				end
+		end,
 
-	%% Check solution seed
+	%% Check solution seed.
 	#nonce_limiter_info{ next_seed = PrevNextSeed,
 			next_vdf_difficulty = PrevNextVDFDifficulty,
 			global_step_number = PrevStepNumber } = TipNonceLimiterInfo,
 	PrevIntervalNumber = PrevStepNumber div ?NONCE_LIMITER_RESET_FREQUENCY,
-	PassesSeedCheck = PassesTimelineCheck andalso
-			{IntervalNumber, NonceLimiterNextSeed, NonceLimiterNextVDFDifficulty}
-					== {PrevIntervalNumber, PrevNextSeed, PrevNextVDFDifficulty},
+	PassesSeedCheck =
+		case PassesTimelineCheck of
+			{false, Reason} ->
+				{false, Reason};
+			true ->
+				case {IntervalNumber, NonceLimiterNextSeed, NonceLimiterNextVDFDifficulty}
+						== {PrevIntervalNumber, PrevNextSeed, PrevNextVDFDifficulty} of
+					false ->
+						ar_events:send(solution, {stale, #{ source => Source }}),
+						{false, seed_data};
+					true ->
+						true
+				end
+		end,
+
+	%% Check solution difficulty
+	Diff = get_current_diff(Timestamp),
+	PassesDiffCheck =
+		case PassesSeedCheck of
+			{false, Reason2} ->
+				{false, Reason2};
+			true ->
+				case binary:decode_unsigned(SolutionH, big) > Diff of
+					false ->
+						ar_events:send(solution, {processed, #{ source => Source }}),
```

Comment:
what does `processed` refer to here?

Reply:
Could we use `partial` instead? So the list of different solution states would be:
`rejected`: invalid solution for some reason
`stale`: valid solution that is now too old
`partial`: valid solution that does not pass the network difficulty check
`accepted`: valid solution that passes the network difficulty check and is not too old